File: grokking-the-coding-interview_patterns-for-coding-questions/3.-pattern-sliding-window/4-fruits-into-baskets-medium/solution_4-fruits-into-baskets-medium.js
const totalFruit = (fruits) => {
    let largest = 0;
    let fMap = {};
    let start = 0;

    fruits.forEach((f, end) => {
        if (!fMap[f]) {
            fMap[f] = 0
        }
        fMap[f]++;
        while (Object.keys(fMap).length > 2) {
            const firstIn = fruits[start];
            fMap[firstIn]--
            if (fMap[firstIn] === 0) {
                delete fMap[firstIn]
            }
            start ++
        }
        largest = Math.max(largest, end - start + 1);
    });
    return largest;
}

console.log(totalFruit([1,2,1]) === 3)
console.log(totalFruit([0,1,2,2]) === 3)
console.log(totalFruit([1,2,3,2,2]) === 4)
console.log(totalFruit([3, 3, 3, 1, 2, 1, 1, 2, 3, 3, 4]) === 5)

File: grokking-the-coding-interview_patterns-for-coding-questions/3.-pattern-sliding-window/4-fruits-into-baskets-medium/solution_4-fruits-into-baskets-medium.py
def total_fruit(fruits):
    fruit_map = {}
    max_length = 0
    start = 0

    for end, fruit in enumerate(fruits):
        if fruit not in fruit_map:
            fruit_map[fruit] = 0
        fruit_map[fruit] += 1
        while len(fruit_map.keys()) > 2:
            fruit_map[fruits[start]] -= 1
            if fruit_map[fruits[start]] == 0:
                del fruit_map[fruits[start]]
            start += 1
        curr_window = end - start + 1
        max_length = curr_window if curr_window > max_length else max_length
    return max_length

print(total_fruit([1,2,1]) == 3)
print(total_fruit([0,1,2,2]) == 3)
print(total_fruit([1,2,3,2,2]) == 4)
print(total_fruit([3, 3, 3, 1, 2, 1, 1, 2, 3, 3, 4]) == 5)
